probe-rs/src/bin/probe-rs/cmd/test.rs

```diff
@@ -0,0 +1,617 @@
+    HaltReason, Lister, MemoryInterface, SemihostingCommand, Session, VectorCatchCondition,
+};
+use probe_rs_target::MemoryRegion;
+use signal_hook::consts::signal;
+use static_cell::StaticCell;
+use time::UtcOffset;
+
+use crate::util::common_options::{BinaryDownloadOptions, ProbeOptions};
+use crate::util::flash::run_flash_download;
+use crate::util::rtt::{self, RttConfig};
+use crate::FormatOptions;
+
+const RTT_RETRIES: usize = 10;
+
+#[derive(clap::Parser)]
+pub struct Cmd {
+    #[clap(flatten)]
+    pub(crate) probe_options: ProbeOptions,
+
+    #[clap(flatten)]
+    pub(crate) download_options: BinaryDownloadOptions,
+
+    /// Always print the stacktrace on ctrl + c.
+    #[clap(long)]
+    pub(crate) always_print_stacktrace: bool,
+
+    /// Whether to erase the entire chip before downloading
+    #[clap(long)]
+    pub(crate) chip_erase: bool,
+
+    /// Suppress filename and line number information from the rtt log
+    #[clap(long)]
+    pub(crate) no_location: bool,
+
+    #[clap(flatten)]
+    pub(crate) format_options: FormatOptions,
+
+    #[clap(long)]
+    pub(crate) log_format: Option<String>,
+
+    /// Enable reset vector catch if its supported on the target.
+    #[arg(long)]
+    pub catch_reset: bool,
+    /// Enable hardfault vector catch if its supported on the target.
+    #[arg(long)]
+    pub catch_hardfault: bool,
+
+    /// Scan the memory to find the RTT control block
+    #[clap(long)]
+    pub(crate) rtt_scan_memory: bool,
+
+    /// <elf> and the remaining arguments for the test runner (list tests, filter tests etc). Run `probe-rs test -- --help` for more information.
+    #[clap(last(true))]
+    pub(crate) libtest_args: Vec<String>,
+}
+
+static SESSION: StaticCell<Session> = StaticCell::new();
+static RUNNER: StaticCell<RefCell<Runner>> = StaticCell::new();
+struct Runner {
+    core: Core<'static>,
+    timestamp_offset: UtcOffset,
+    path: String,
+    always_print_stacktrace: bool,
+    no_location: bool,
+    log_format: Option<String>,
+    memory_map: Vec<MemoryRegion>,
+    rtt_scan_regions: Vec<Range<u64>>,
+}
+
+const SEMIHOSTING_USER_LIST: u32 = 0x100;
+
+impl Cmd {
+    pub fn run(
+        self,
+        lister: &Lister,
+        run_download: bool,
+        timestamp_offset: UtcOffset,
+    ) -> Result<()> {
+        let path = self.libtest_args[0].clone();
+        let libtest_args = libtest_mimic::Arguments::from_iter(self.libtest_args);
+
+        let (mut session, probe_options) = self.probe_options.simple_attach(lister)?;
+
+        if run_download {
+            let mut file = match File::open(&path) {
+                Ok(file) => file,
+                Err(e) => {
+                    return Err(FileDownloadError::IO(e)).context("Failed to open binary file.")
+                }
+            };
+
+            let mut loader = session.target().flash_loader();
+
+            let format = self.format_options.into_format(session.target())?;
+            match format {
```

Comment:
Have a look at #1973 for how to simplify this :)

Reply:
Done